ergo-lib/src/wallet/ext_secret_key.rs

```diff
@@ -0,0 +1,267 @@
+    sigma_protocol::sigma_boolean::ProveDlog,
+};
+use hmac::{Hmac, Mac, NewMac};
+use sha2::Sha512;
+use thiserror::Error;
+
+/// Private key (serialized Scalar) bytes
+pub type SecretKeyBytes = [u8; 32];
+/// Chain code bytes
+pub type ChainCode = [u8; 32];
+
+type HmacSha512 = Hmac<Sha512>;
+
+/// Result type for ExtendedSecretKey operations
+pub type Result<T> = std::result::Result<T, ExtSecretKeyError>;
+
+/// Extended secret key
+/// implemented according to BIP-32
+#[derive(PartialEq, Debug, Clone)]
+pub struct ExtSecretKey {
+    /// The secret key
+    private_input: DlogProverInput,
+    chain_code: ChainCode,
+    derivation_path: DerivationPath,
+}
+
+/// Extended secret key errors
+#[derive(Error, PartialEq, Eq, Debug, Clone)]
+pub enum ExtSecretKeyError {
+    /// Parsing error
+    #[error("parsing error: {0}")]
+    SigmaParsingError(#[from] SigmaParsingError),
+    #[error("serialization error: {0}")]
+    /// Serializing error
+    SigmaSerializationError(#[from] SigmaSerializationError),
+    /// Error encoding bytes as SEC-1-encoded scalar
+    #[error("scalar encoding error")]
+    ScalarEncodingError,
+    /// Derivation path child index error
+    /// For example trying to use a u32 value for a private index (31 bit size)
+    #[error("child index error: {0}")]
+    ChildIndexError(#[from] ChildIndexError),
+}
+
+impl ExtSecretKey {
+    const BITCOIN_SEED: &'static [u8; 12] = b"Bitcoin seed";
+
+    /// Create a new extended secret key instance
+    pub fn new(
+        secret_key_bytes: SecretKeyBytes,
+        chain_code: ChainCode,
+        derivation_path: DerivationPath,
+    ) -> Result<Self> {
+        let private_input = DlogProverInput::from_bytes(&secret_key_bytes)
+            .ok_or(ExtSecretKeyError::ScalarEncodingError)?;
+        Ok(Self {
+            private_input,
+            chain_code,
+            derivation_path,
+        })
+    }
+
+    /// Byte representation of the underlying scalar
+    pub fn secret_key_bytes(&self) -> SecretKeyBytes {
+        self.private_input.to_bytes()
+    }
+
+    /// Public image associated with the private input
+    pub fn public_image(&self) -> ProveDlog {
+        self.private_input.public_image()
+    }
+
+    /// Public image bytes in SEC-1 encoded & compressed format
+    pub fn public_image_bytes(&self) -> Result<Vec<u8>> {
+        Ok(self.public_image().h.sigma_serialize_bytes()?)
+    }
+
+    /// The extended public key associated with this secret key
+    #[allow(clippy::unwrap_used)]
+    pub fn public_key(&self) -> Result<ExtPubKey> {
+        Ok(ExtPubKey::new(
+            // unwrap is safe as it is used on an Infallible result type
+            self.public_image_bytes()?.try_into().unwrap(),
+            self.chain_code,
+            self.derivation_path.clone(),
+        )?)
+    }
+
+    /// Derive a child extended secret key using the provided index
+    #[allow(clippy::unwrap_used)]
```

Comment:
I prefer keeping `allow` directive above the actual call. It saves you hunting for it inside the function and disincentivizes blanket `allow`.